packages/alpinejs/src/directives/x-for.js

```diff
@@ -11,226 +11,133 @@ directive('for', (el, { expression }, { effect, cleanup }) => {
-            }
-        }
-
-        // Rather than making DOM manipulations inside one large loop, we'll
-        // instead track which mutations need to be made in the following
-        // arrays. After we're finished, we can batch them at the end.
-        let adds = []
-        let moves = []
-        let removes = []
-        let sames = []
-
-        // First, we track elements that will need to be removed.
-        for (let i = 0; i < prevKeys.length; i++) {
-            let key = prevKeys[i]
-
-            if (keys.indexOf(key) === -1) removes.push(key)
-        }
-
-        // Notice we're mutating prevKeys as we go. This makes it
-        // so that we can efficiently make incremental comparisons.
-        prevKeys = prevKeys.filter(key => ! removes.includes(key))
-
-        let lastKey = 'template'
-
-        // This is the important part of the diffing algo. Identifying
-        // which keys (future DOM elements) are new, which ones have
-        // or haven't moved (noting where they moved to / from).
-        for (let i = 0; i < keys.length; i++) {
-            let key = keys[i]
-
-            let prevIndex = prevKeys.indexOf(key)
-
-            if (prevIndex === -1) {
-                // New key found.
-                prevKeys.splice(i, 0, key)
-
-                adds.push([lastKey, i])
-            } else if (prevIndex !== i) {
-                // A key has moved.
-                let keyInSpot = prevKeys.splice(i, 1)[0]
-                let keyForSpot = prevKeys.splice(prevIndex - 1, 1)[0]
-
-                prevKeys.splice(i, 0, keyForSpot)
-                prevKeys.splice(prevIndex, 0, keyInSpot)
-
-                moves.push([keyInSpot, keyForSpot])
-            } else {
-                // This key hasn't moved, but we'll still keep track
-                // so that we can refresh it later on.
-                sames.push(key)
-            }
-
-            lastKey = key
-        }
-
-        // Now that we've done the diffing work, we can apply the mutations
-        // in batches for both separating types work and optimizing
-        // for browser performance.
-
-        // We'll remove all the nodes that need to be removed,
-        // and clean up any side effects they had.
-        for (let i = 0; i < removes.length; i++) {
-            let key = removes[i]
-
-            if (! (key in lookup)) continue
-
-            mutateDom(() => {
-                destroyTree(lookup[key])
+        /**
+         * In order to remove Elements early we need to generate the key/scope pairs
+         * up front, moving existing elements from the old lookup and to the new.
+         * This leaves only the Elements to be removed in the old lookup.
+         */
+        const oldLookup = templateEl._x_lookup
+        const lookup = new Map()
+        templateEl._x_lookup = lookup
+        const scopeEntries = []
+        const hasStringKeys = isObject(items)
+        Object.entries(items).forEach(([index, item]) => {
```

Comment:
Using `Object.keys` and then doing `items[key]` is more verbose but may be faster and more memory efficient, though I am not sure.

Also the forEach could be replaced with a standard `for` loop which is faster too.